### PR TITLE
Make /sync "better".

### PR DESCRIPTION
**Pull request by erikjohnston**
Things:

- Add cache that allows us to work out what rooms have changed since a `stream` token.
- Fetch new events for each room separately, relying on the above cache to remove unnecessary queries.
- Amalgamate the different paths that produce the `JoinSyncResult`
- Allow pagination API to work with `stream` tokens.

TODO:

- [x] Add index `on events (room_id, stream_ordering)`

### CODE REVIEW

**Comment by erikjohnston** [comment]
See https://github.com/matrix-org/sytest/pull/164 for some changes to sytest.

File: synapse/util/caches/room_change_cache.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    """
+    def __init__(self, name, current_key, size_of_cache=10000):
+        self._size_of_cache = size_of_cache
+        self._room_to_key = {}
+        self._cache = sorteddict()
+        self._earliest_known_key = current_key
+        self.name = name
+        caches_by_name[self.name] = self._cache
+
+    def get_room_has_changed(self, room_id, key):
```
Comment (NegativeMjark) [inline synapse/util/caches/room_change_cache.py]: Some docstring would be nice here.

**Comment by NegativeMjark** [comment]
Apart from the missing index needed for get_max_topological_token_for_stream_and_room it LGTM

**Comment by erikjohnston** [comment]
matrixbot: test this please

**Comment by erikjohnston** [comment]
@NegativeMjark PTAL

The only things that have really changed are the `StreamChangeCache` and the fact that we now also cache account data and tags.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**